### PR TITLE
new feature: options.routeFound

### PR DESCRIPTION
**Pull request by Sly777**
Added option to handle process after route found

Example: 

```javascript
const routeHandler = routes.getRequestHandler(app, {
  routeFound: (req, res, url, queryParams) => {
    renderAndCache(req, res, url, queryParams);
  }
});

function renderAndCache() {
   // handle for caching
}
```

### CODE REVIEW

**Comment by fridays** [comment]
Thanks for contributing! Please see if https://github.com/fridays/next-routes/pull/51 works for you

**Comment by Sly777** [comment]
I'm closing this PR because of #51 . Thanks @fridays!

**Comment by fridays** [comment]
Reopening because it is the right place for a request handler 🌴

File: src/index.js [hunk 1]
```diff
@@ -37,16 +40,21 @@ class Routes {
     return {route, params}
   }
 
-  getRequestHandler (app) {
+  getRequestHandler (app, options = {}) {
```
Comment (fridays) [inline src/index.js]: I suggest remove options for now and take the function directly as `handler`.

File: src/index.js [hunk 2]
```diff
@@ -37,16 +40,21 @@ class Routes {
...
     const nextHandler = app.getRequestHandler()
+    this.options = Object.assign({}, this.options, options)
```
Comment (fridays) [inline src/index.js]: This can be removed

File: src/index.js [hunk 3]
```diff
@@ -37,16 +40,21 @@ class Routes {
...
 
     return (req, res) => {
       const parsedUrl = parse(req.url, true)
       const {pathname, query} = parsedUrl
       const {route, params} = this.match(pathname)
 
       if (route) {
-        app.render(req, res, route.page, {...query, ...params})
+        if (this.options.routeFound && typeof this.options.routeFound === 'function') {
```
Comment (fridays) [inline src/index.js]: What do you think of:

```javascript
if (route) {
  Object.assign(query, params)

  if (handler) {
    handler({req, res, route, query})
  } else {
    app.render(req, res, route.page, query)
  }
}
```
The handler gets an object as single parameter including the full route object

**Comment by Sly777** [comment]
Hey @fridays, I made changes as you said but the only difference is I chose a name as "customHandler" to avoid some wording confusion.

**Comment by fridays** [comment]
Thanks, looks good!

File: README.md [hunk 4]
```diff
@@ -69,6 +69,12 @@ app.prepare().then(() => {
 })
 
 ```
+Optionally you can pass a custom handler, for example:
+```javascript
+const handler = routes.getRequestHandler(app, ({req, res, route, query}) => {
+  renderAndCache(req, res, route.page, query)
```
Comment (HaNdTriX) [inline README.md]: I recommend removing `renderAndCache` and instead show how to properly handle the route manually. (```app.render(...)```)

**Comment by Sly777** [comment]
I updated the example @HaNdTriX

**Comment by fridays** [comment]
@Sly777 I think @HaNdTriX meant to only change `renderAndCache` to `app.render` in the example. Query transforms are not implemented here and redirects should better happen inside getInitialProps. If you could change this, we can merge 👍

**Comment by Sly777** [comment]
oooh, I see :) Okay I updated as you want, @fridays.

**Comment by fridays** [comment]
Thanks! It's now there in version `1.0.29`

**Pull request merged**